File: src/main/java/com/munzenberger/feed/handler/DownloadEnclosures.java
/*
 * Copyright 2020 Brian Munzenberger
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * 		http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.munzenberger.feed.handler;

import com.munzenberger.feed.log.Logger;
import com.munzenberger.feed.parser.rss.Enclosure;
import com.munzenberger.feed.parser.rss.Item;
import com.munzenberger.feed.util.DataTransfer;
import com.munzenberger.feed.util.DateParser;
import com.munzenberger.feed.util.Formatter;
import com.munzenberger.feed.util.URLProcessor;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLDecoder;
import java.util.Date;

public class DownloadEnclosures implements ItemHandler {

	private String targetDir = ".";

	private String filter = null;

	@SuppressWarnings("unused")
	public void setTargetDir(String targetDir) {
		this.targetDir = targetDir;
	}

	public void setFilter(String filter) {
		this.filter = filter;
	}

	@Override
	public void process(Item item, Logger logger) throws ItemHandlerException {
		for (Enclosure e : item.getEnclosures()) {
			process(e, item.getPubDate(), logger);
		}
	}

	protected void process(Enclosure e, String pubDate, Logger logger) throws ItemHandlerException {

		if (evaluate(filter, e)) {

			File file = process(e.getUrl(), logger);

			Date date = DateParser.parse(pubDate, logger);
			if (date != null) {
				if (!file.setLastModified(date.getTime())) {
					logger.log("Could not set last modified time on file: " + file);
				}
			}
		}
	}

	protected File process(String downloadURL, Logger logger) throws ItemHandlerException {

		URL url;

		try {
			url = new URL(downloadURL);
		} catch (MalformedURLException ex) {
			throw new ItemHandlerException(ex);
		}

		File file = getLocalFile(url);

		try {
			download(url, file, logger);
		}
		catch (IOException ex) {
			if (!file.delete()) {
				logger.log("Could not delete file: " + file);
			}
			throw new ItemHandlerException(ex);
		}

		return file;
	}

	public File getLocalFile(URL url) throws ItemHandlerException {

		String file = url.getPath();

		int i = file.lastIndexOf('/');
		if (i >= 0) {
			file = file.substring(i+1);
		}

		try {
			file = URLDecoder.decode(file, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			throw new ItemHandlerException(e);
		}

		String ext = "";
		i = file.lastIndexOf('.');
		if (i > 0) {
			ext = file.substring(i);
			file = file.substring(0, i);
		}

		file = Formatter.fileName(file);
		ext = Formatter.fileName(ext);

		String localPath = targetDir + File.separator + file + ext;
		File localFile = new File(localPath);

		if (localFile.exists()) {
			// append a timestamp to make this file unique
			localPath = targetDir + File.separator + file + "-" + System.currentTimeMillis() + ext;
			localFile = new File(localPath);

			if (localFile.exists()) {
				throw new ItemHandlerException("File already exists: " + file);
			}
		}

		return localFile;
	}

	protected void download(URL url, File file, Logger logger) throws IOException {
		long time = System.currentTimeMillis();		
		long bytes;

		logger.log("Transferring " + url + " -> " + file);

		InputStream in = URLProcessor.getResponse(url).getInputStream();
		OutputStream out = new FileOutputStream(file);

		bytes = DataTransfer.transfer(in, out);

		in.close();
		out.flush();
		out.close();

		time = System.currentTimeMillis() - time;
		logger.log(Formatter.fileSize(bytes) + " transferred in " + Formatter.elapsedTime(time));
	}

	protected boolean evaluate(String filter, Enclosure enclosure) {

		boolean eval = true;

		if (filter != null) {
			eval = enclosure.getUrl().matches(filter);
		}

		return eval;
	}
}

File: src/main/kotlin/com/munzenberger/feed/handler/DownloadEnclosures.kt
package com.munzenberger.feed.handler

import com.munzenberger.feed.FeedContext
import com.munzenberger.feed.Item
import com.munzenberger.feed.URLClient
import com.munzenberger.feed.config.filteredForPath
import com.munzenberger.feed.filename
import okio.buffer
import okio.sink
import okio.source
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.net.URL
import java.net.URLDecoder
import java.text.NumberFormat

class DownloadEnclosures : ItemHandler {

    var targetDirectory: String = "."

    override fun execute(context: FeedContext, item: Item) {
        item.enclosures.forEach { enclosure ->

            print("Resolving enclosure source... ")

            URLClient.connect(URL(enclosure.url)).run {

                println(resolvedUrl)

                contentDisposition?.let {
                    println("Content-Disposition: $it")
                }

                val target = targetFileFor(filename)

                print("Downloading to $target... ")

                val result = profile { download(inStream, target) }

                println("${result.first.formatAsSize()} transferred in ${result.second.formatAsTime()}.")

                item.timestampAsInstant?.let {
                    if (!target.setLastModified(it.toEpochMilli())) {
                        System.err.println("Could not set last modified time on file: $target")
                    }
                }
            }
        }
    }

    internal fun targetFileFor(filename: String): File {

        var path = targetDirectory + File.separator + filename

        var targetFile = File(path)

        if (targetFile.exists()) {

            // insert a timestamp into the filename to make it unique

            var name = path
            var extension = ""

            val i = path.lastIndexOf('.')
            if (i > 0) {
                name = path.substring(0, i)
                extension = path.substring(i)
            }

            path = name + "-" + System.currentTimeMillis() + extension

            targetFile = File(path)

            if (targetFile.exists()) {
                throw IOException("Local file already exists: $targetFile")
            }
        }

        return targetFile
    }
}

internal val URLClient.Response.filename: String
    // use the filename from the content disposition header, if present
    get() = contentDisposition.filename?.filteredForPath() ?: resolvedUrl.filename

internal val URL.filename: String
    get() = this.path.urlDecode().split('/').last()

private fun String.urlDecode(encoding: String = "UTF-8"): String {
    // handles nested URLs
    return when (val decoded = URLDecoder.decode(this, encoding)) {
        this -> this
        else -> decoded.urlDecode(encoding)
    }
}

private fun download(inStream: InputStream, target: File): Long {

    val inputSource = inStream.source().buffer()
    val outputSink = target.sink().buffer()

    return inputSource.use { input ->
        outputSink.use { output ->
            input.readAll(output)
        }
    }
}

private fun <T> profile(block: () -> T): Pair<T, Long> {
    val start = System.currentTimeMillis()
    val v = block.invoke()
    val time = System.currentTimeMillis() - start
    return v to time
}

fun Long.formatAsTime(): String {

    val seconds = this / 1000 % 60
    val minutes = this / 1000 / 60

    return when {
        minutes > 0 -> "$minutes min $seconds sec"
        seconds > 0 -> "$seconds sec"
        else -> "$this ms"
    }
}

fun Long.formatAsSize(): String {

    val format = NumberFormat.getInstance().apply {
        minimumFractionDigits = 1
        maximumFractionDigits = 2
    }

    val mbs = this.toDouble() / 1024.0 / 1024.0

    return format.format(mbs) + " MB"
}